awkward/array/jagged.py

```diff
@@ -1350,9 +1362,42 @@ def concatenate(isclassmethod, cls_or_self, arrays):
 
         return cls(starts, stops, content)
 
-    @classmethod
-    def stack(cls, first, *rest):    # each item in first followed by second, etc.
-        raise NotImplementedError
+    @awkward.util.bothmethod
+    def stack(isclassmethod, cls_or_self, arrays):    # each item in first followed by second, etc.
+
+        if isclassmethod: 
+            cls = cls_or_self
+            if not all(isinstance(x, JaggedArray) for x in arrays):
+                raise TypeError("cannot stack non-JaggedArrays with JaggedArray.stack")
+        else:
+            self = cls_or_self
+            cls = self.__class__
+            if not isinstance(self, JaggedArray) or not all(isinstance(x, JaggedArray) for x in arrays):
+                raise TypeError("cannot stack non-JaggedArrays with JaggedArray.stack")
+            arrays = (self,) + tuple(arrays)
+
+        if len(set([len(a) for a in arrays])) > 1:
+            raise ValueError("cannot stack JaggedArrays of different lengths")
+
+        np = cls.numpy
+
+        n = len(arrays)
+        n_content = sum([len(a.flatten()) for a in arrays])
+        dtype = arrays[0].content.dtype
```

Comment:
What if the first array has `dtype=int` and the others have `dtype=float`? You don't want floating point numbers to turn into integers; you want the most general type. If you've already guaranteed that the `arrays` have `numpy.ndarray` contents (see comment on line 1400), then you could add the first flattened element of each:

```python
dtype = np.dtype(sum(x[0] for x in flatarrays if len(x) != 0, False))
```

Summing a single element is not too expensive and it implements the right type-promotion from boolean to int to float to complex. Note that we start the sum with a boolean zero: `False`. That way, an array of booleans won't be unintentionally promoted to integers.

Reply:
Hmm when I try this idea out, for example with `type([True, True], False)`, it is unfortunately an int. It seems to me this shortcut via the tentative `sum` does not work if there are only booleans. Or am I understanding something wrong?

Thanks a lot for all the feedback!

Reply:
Oh, then maybe you just need to handle boolean as a special case. (If it's adding `True + True`, that's like `2`, which is too high of a number to be boolean...) If all arrays are boolean, the result is boolean, and otherwise addition will work.

Here's another way that takes advantage of some logic directly intended to promote types (that we control):

```python
cls.awkward.UnionArray([], [], arrays).dtype
```

And this doesn't like to merge booleans with numbers, either (it returns Numpy's object type, which we should avoid whenever possible).

Reply:
Ok! I went with handling as a special case now and made an extra unit test for it.